core/src/main/java/org/bitcoinj/base/internal/Buffers.java

```diff
@@ -54,10 +56,12 @@ public static byte[] readBytes(ByteBuffer buf, int length) throws BufferUnderflo
      * @param buf buffer to read from
      * @return read bytes
      * @throws BufferUnderflowException if the read value extends beyond the remaining bytes of the buffer
+     * @throws ProtocolException if the length prefix is greater than the maximum allowed message size
      */
-    public static byte[] readLengthPrefixedBytes(ByteBuffer buf) throws BufferUnderflowException {
-        int length = VarInt.read(buf).intValue();
-        return readBytes(buf, length);
+    public static byte[] readLengthPrefixedBytes(ByteBuffer buf) throws BufferUnderflowException, ProtocolException {
+        long length = VarInt.read(buf).longValue();
+        check(length < Message.MAX_SIZE, () -> new ProtocolException("length prefix too long"));
```

Comment:
I think we should check against `Integer.MAX_VALUE` here. This method should not know about Message. And if this is used from message parsing, the payload cannot exceed `Message.MAX_SIZE` anyway.

We could perhaps make the `Integer.MAX_VALUE` check a method of `VarInt`, somehting like `boolean fitsInt()`. Then there would be no need to convert to long and then back to int.

Reply:
Also I think we should continue to throw `BufferunderflowException`, not `ProtocolException`. Because in fact it's exactly that.